cpg-infrastructure/developer.rst

```diff
@@ -0,0 +1,142 @@
+    - Contains the Pulumi configuration for the production environment.
+    - Basically no extra information, except a pointer to the config.
+
+- ``datasets/$DATASET/*``
+    - ``budgets.yaml``
+        - Contains the budgets for the dataset.
+        - See [Budgets](../budgets.md) for more information.
+    - ``members.yaml``
+        - A map of a dataset group, to a list of users.
+        - See the ``setup_externally_specified_members`` method for it's implementation.
+    - ``repositories.yaml``:
+        - A list of repositories for which you can run code against within the analysis-runner for this dataset. (Note, there are default repositories).
+        - This is implemented in ``tokens/main.py`` in cpg-infra-private.
+
+Group membership
+----------------
+
+Groups are used to control access to resources and services. In most places, we add a group to a cloud resource, and let the cloud provider handle the rest. In other cases (like the web service / metamist / analysis-runner), we can't just lookup the group membership from the cloud provider (because it's slow, and it's a little tedious to add members not defined within that cloud), so we have a members cache (a file that stores the group members).
+
+We bring all group memberships under cpg-infrastructure, this allows us to:
+
+- unwrap groups completely, hence we know the full list of users / accounts that need to access specific resources.
+- better version control and store history of group memberships. In the CPG's configuration, we have CODEOWNER policies to manage who can add / remove users from groups.
+
+
+In code, we have a wrapper around a group, which is a list of GroupMember's, which is either a user or itself another group. We track which cloud we need to create the group for. At anytime you can request a full list of users, however the ID component might be a pulumi output value, eg: you can't know the service-account ID before you create it.
+
+All dataset infrastructure methods can create groups, and add members to them, and at the very end, these group members are finalised, synced to the cloud resource, and any access group caches are updated.
+
+
+How does it actually get deployed?
+----------------------------------
+
+
+In cpg-infrastructure-private, we have a deploy.yaml GitHub action. This:
+
+- Installs the latest cpg-infrastructure package
+- Runs the pulumi up command:
+    - This internally runs the ``__main__.py`` file, which is the entry point for the pulumi program. This packages up all the configuration in the private repo into:
+
+        1. A ``CPGInfrastructureConfig`` object, which contains users.
+        2. A ``list[CPGDatasetConfig]`` objects, which contains the datasets to spin up.
+
+Puluim does the rest!
+
+
+Previewing the workflow locally
+-------------------------------
+
+Prerequisites:
+
+* Have installed:
+    * Pulumi
+    * google-cloud-sdk
+    * azure-cli
+* Get and export the pulumi access token from a CPG team-member
```

Comment:
Is this referring to `PULUMI_CONFIG_PASSPHRASE` ? If so, it might be better to refer to it as that, I wouldn't be quite sure what the pulumi access token is talking about